### ISSUE TITLE
[GitHub actions] KeyError: 'service_job_id'

### ISSUE THREAD

**Issue by escaped** [coveralls-clients/coveralls-python#247]
Hey,

thanks for this very useful client. I use it in almost all of my open source projects, but it recently stopped working. At first, I thought it was related to the new `2.2.0` release, but since I ran an older (previously successful) pipeline using `2.1.2` again and it failed, I think the issue lies somewhere else.

I use GitHub actions and integrated it like this ([full workflow](https://github.com/escaped/networkx-astar-path/blob/95766bb3845629363872494ff10a51e22b8ca1ab/.github/workflows/test.yml)):
```yaml
...
test:
      ....
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install tox tox-gh-actions coveralls  # or coveralls==2.1.2
      ...
      - name: coveralls
        run: coveralls  # or `coveralls debug` for debugging
        env:
          COVERALLS_PARALLEL: true
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
...
  coveralls:
    needs: [test]
    runs-on: ubuntu-latest
    steps:
      - name: Set up Python 
        uses: actions/setup-python@v2
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install coveralls
      - name: coveralls
        run: coveralls --finish  # or `coveralls debug --finish` for debugging
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

When running the pipeline, this is the error I am getting:

![image](https://user-images.githubusercontent.com/318650/101292056-fabf7800-37c1-11eb-9b5d-33d69f30a26c.png)

To debug this, I created a new [branch](https://github.com/escaped/networkx-astar-path/pull/4) based on a failing branch:
- [Initial run **failed**, version `2.2.0`](https://github.com/escaped/networkx-astar-path/pull/4/checks?check_run_id=1507482203)
- [Using `coverall debug` made the pipeline **pass**, version `2.2.0`](https://github.com/escaped/networkx-astar-path/runs/1507487604)
- [Reverting the commit and the pipeline **fails** again, version `2.2.0`](https://github.com/escaped/networkx-astar-path/pull/4/checks?check_run_id=1507493454)
- [Switching to version `2.1.2` **fails** the pipeline as well](https://github.com/escaped/networkx-astar-path/pull/4/checks?check_run_id=1507499152)
- [Using `coveralls debug` with version `2.1.2` **works** again](https://github.com/escaped/networkx-astar-path/pull/4/checks?check_run_id=1507506515)

I am a bit lost now; looks like it working fine when using the `debug` mode, but doesn't without it. Does anybody have an idea if that's a bug or a configuration issue on my side?

**Comment by richm** [comment]
The payload sent to coveralls.io from a Github Actions job must contain `{"service_job_id":null, "service_name": "github"}`. By default, the coveralls library omits `service_job_id` and uses `github-actions`.

**Comment by richm** [comment]
I worked around the issue like this: https://github.com/linux-system-roles/tox-lsr/pull/7

**Comment by goerz** [comment]
I'm running into the same problem. It's because the service_job_id never get set here:

https://github.com/coveralls-clients/coveralls-python/blob/a9b36299ce9ba3bb6858700781881029d82e545d/coveralls/api.py#L56-L57

because `jobs` is `None` from

https://github.com/coveralls-clients/coveralls-python/blob/a9b36299ce9ba3bb6858700781881029d82e545d/coveralls/api.py#L109

which then blows up in

https://github.com/coveralls-clients/coveralls-python/blob/a9b36299ce9ba3bb6858700781881029d82e545d/coveralls/api.py#L224-L225

Can't we just change that to the following?
~~~
new_id = '{}-{}'.format(
    self.config.get('service_job_id', None), random.randint(0, sys.maxsize))
~~~

**Comment by J-M0** [comment]
I found a weird way to fix this. I initially set up my workflow with the COVERALLS_FLAG_NAME variable set and got this error. Even after I removed that from the workflow, I would still get this error. I then deleted my repo on coveralls.io and reran the jobs without the flag name and they succeeded. Links here:

1. [Initial run with COVERALLS_FLAG_NAME set](https://github.com/J-M0/ci-experiments/actions/runs/426152052)
2. [Run without COVERALLS_FLAG_NAME](https://github.com/J-M0/ci-experiments/actions/runs/426167333)

**Comment by Silmathoron** [comment]
I'm encountering the same issue in a different context and I'm afraid the above discussion did not really clarify the problem for me.
I don't know whether my issue is related or if I should open a new issue, so I'll first expose the problem here.

I used to use Travis to run tests but I'm now moving to [SourceHut](https://sourcehut.org/), I configured the new setting (using repo token etc) and everything seems to run fine except for that ``service_job_id`` error...
Does this mean additional variables should be set in ``.coveralls.yml`` or as environment variables?
I did not see any change setting ``COVERALLS_FLAG_NAME`` to a specific value but debug indeed works fine.
So far I only set

    service_name: sourcehut
    repo_token: mytoken

**Comment by TheKevJames** [comment]
@escaped looks like your original issue here is related to what @richm suggested -- setting the service_job_id to null and the service_name to github-actions should hopefully fix up your usecase. I am in progress releasing `coveralls==3.0.0` which will fix these issues; please take a look and let me know if that fixes your issue!

Note that v3 will change the precedence order for config files -- user-provided configuration will now *override* the CI values. If you are still having issues with v3 you can, for example, try setting `export COVERALLS_SERVICE_NAME=github` to override the value we're sending (we send `github-actions` by default). If I'm reading your issue correctly, the default behaviour in v3 _should_ work for you? To be honest, the Github Actions API seems to keep changing under me, or at least I'm not understanding what it's expecting properly, so YMMV.

With respect to the missing service_job_id key, it's worth noting that only occurs during the retry -- you shouldn't need to retry at all once the fix is in place. That said, thanks for pointing it out; I'll add a fix for that as well!

@richm thanks for the instructions and the link to your PR! You should be able to make use of the config values in v3 to avoid needing that custom script, if you'd like -- just set `COVERALLS_SERVICE_NAME=github` (or use the `.coveralls.yml` or CLI flags, if you'd prefer those).

@goerz @J-M0 @Silmathoron I suspect the issues y'all are running into will also be solved with these v3 fixes.
